### PR TITLE
Add service tokens

### PR DESCRIPTION
**Pull request by koszti**
Add option to generate Service Tokens to programatically access API endpoints using secure [JWT](https://jwt.io/) based authentication. Service Tokens meant to use for service-to-service communication, integration and automation of different services. Service tokens don't required pre-authentication to SqlPad, the credentials and permissions embedded in the token and it can communicate with the SqlPad API using single HTTP request.

Screenshot 2020-03-23 at 00 30 38

The service tokens are [JSON Web Tokens](https://jwt.io/) that can be generated in the UI. We can specify, **role** and **expiry** date for the token to keep every endpoint "role based" and secure even if it's accessed by services.

### CODE REVIEW

**Comment by koszti** [comment]
sorry, this is work in progress and I sent here accidentally. Please feel free to close it or take it only as sneak peek for now. Tests and resolving conficts in progress and I will sequelize the model. Thanks for the tip!

**Comment by koszti** [comment]
hey @rickbergfalk thanks for the comments.

I changed it to sqlite/sequelize, and added server tests. Actually I really like moving to sequelize because we can use external RDMSs as the backend db that runs outside of the container. Great!

The test cases are focusing on generating/querying/deleting tokens and not testing using them. Can you please suggest or help with adding authentication tests if you have something in progress? Otherwise this PR is now ready to merge.

**Comment by rickbergfalk** [comment]
@koszti glad you like the move to sequelize! 

I'll take a look and review this later today. 

I still need to document it, but I'm mostly wrapped on my proxy auth work. For signup/signin local auth tests, and proxy auth tests, I'm starting to use the `supertest` library directly, with a `TestUtils` that has a specific config set up for whatever the config is being tested. https://github.com/rickbergfalk/sqlpad/pull/588

Hopefully the examples help. Otherwise I'm okay adding tests separately after this is merged too

File: server/middleware/must-be-authenticated.js [hunk 1]
```diff
@@ -1,7 +1,52 @@
...
+
+// If authenticated continue, otherwise try with JWT service tokens or
+// redirect user to signin
+function mustBeAuthenticated(req, res, next) {
   if (req.isAuthenticated()) {
     return next();
   }
-  res.redirect(req.config.get('baseUrl') + '/signin');
-};
+  if (req.config.get('serviceTokenSecret')) {
```
Comment (rickbergfalk) [inline server/middleware/must-be-authenticated.js]: I like this here - I'll probably move the authProxy stuff here as well once both are merged

File: server/lib/db.js [hunk 2]
```diff
@@ -60,13 +60,15 @@ async function initNedb(config) {
     connectionAccesses: getDatastore('connectionaccesses.db'),
     queries: getDatastore('queries.db'),
     queryHistory: getDatastore('queryhistory.db'),
+    serviceTokens: getDatastore('servicetokens.db'),
```
Comment (rickbergfalk) [inline server/lib/db.js]: I think this was missed with move to sequelize/sqlite

File: server/models/service-tokens.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
+const jwt = require('jsonwebtoken');
+
+// NOTE - because ServiceTokens is driven off of Sequelize ORM model
+// and not nedb (which is schemaless) I am skipping defining a Joi schema here.
+// For info on what QueryAcl schema is, see sequelize/QueryAcl.js
```
Comment (rickbergfalk) [inline server/models/service-tokens.js]: Can you remove this comment or update?

File: server/lib/db.js (hunk 2 above)
Comment (koszti) [reply server/lib/db.js]: fixed by 5096803f85d0d787b0db640c2df59dc0b4f3f5e2

File: server/models/service-tokens.js (hunk 3 above)
Comment (koszti) [reply server/models/service-tokens.js]: fixed by 6fb43c036cfe05c11a2277b95820a30bbb4a1634

**Comment by koszti** [comment]
>  For signup/signin local auth tests, and proxy auth tests, I'm starting to use the `supertest` library directly, with a `TestUtils` that has a specific config set up for whatever the config is being tested.
> 
> Hopefully the examples help.

thanks, it helped. I added some more tests using the `supertest` lib.

**Pull request merged**